rust/azure_iot_operations_services/src/azure_device_registry/client.rs

```diff
@@ -1,2 +1,1559 @@
+    ///
+    /// [`struct@Error`] of kind [`AIOProtocolError`](ErrorKind::AIOProtocolError)
+    /// if device, inbound endpoint are invalid.
+    ///
+    /// [`struct@Error`] of kind [`ObservationError`](ErrorKind::ObservationError)
+    /// if the observation was not successful.
+    pub async fn unobserve_device_update_notifications(
+        &self,
+        device_name: String,
+        inbound_endpoint_name: String,
+        timeout: Duration,
+    ) -> Result<(), Error> {
+        let unobserve_payload =
+            adr_name_gen::SetNotificationPreferenceForDeviceUpdatesRequestPayload {
+                notification_preference_request: adr_name_gen::NotificationPreference::Off,
+            };
+
+        let unobserve_request =
+            adr_name_gen::SetNotificationPreferenceForDeviceUpdatesRequestBuilder::default()
+                .payload(unobserve_payload)
+                .map_err(ErrorKind::from)?
+                .topic_tokens(Self::get_topic_tokens(
+                    device_name.clone(),
+                    inbound_endpoint_name.clone(),
+                ))
+                .timeout(timeout)
+                .build()
+                .map_err(ErrorKind::from)?;
+        let response = self
+            .notify_on_device_update_command_invoker
+            .invoke(unobserve_request)
+            .await
+            .map_err(ErrorKind::from)?;
+        match response.payload.notification_preference_response {
+            adr_name_gen::NotificationPreferenceResponse::Accepted => {
+                let receiver_id = Self::hash_device_endpoint(&device_name, &inbound_endpoint_name);
+                // Remove it from our dispatcher
+                if self
+                    .device_update_notification_dispatcher
+                    .unregister_receiver(&receiver_id)
+                {
+                    log::debug!(
+                        "Device `{device_name:?}` with inbound endpoint `{inbound_endpoint_name:?}` removed from observed list"
+                    );
+                } else {
+                    log::debug!(
+                        "Device `{device_name:?}` with inbound endpoint `{inbound_endpoint_name:?}` not in observed list"
+                    );
+                }
+                Ok(())
+            }
+            adr_name_gen::NotificationPreferenceResponse::Failed => {
+                Err(Error(ErrorKind::ObservationError))
+            }
+        }
+    }
+
+    /// Hashes the device name and inbound endpoint name to create a single string.
+    fn hash_device_endpoint(device_name: &str, inbound_endpoint_name: &str) -> String {
+        // `~`` can't be in a topic token, so this will never collide with another device + inbound endpoint name combo
+        format!("{device_name}~{inbound_endpoint_name}")
+    }
+
+    /// Unhashes the device name and inbound endpoint name from a single string.
+    fn un_hash_device_endpoint(hashed_device_endpoint: &str) -> Option<(String, String)> {
```

Comment:
"unhash" should probably be one word. It's not technically a word, but "un" by itself isn't either